File: common/src/main/java/dev/mayaqq/estrogen/resources/thighhighs/ThighHighStyleLoader.java
package dev.mayaqq.estrogen.resources.thighhighs;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonParser;
import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Codec;
import com.mojang.serialization.JsonOps;
import dev.mayaqq.estrogen.Estrogen;
import dev.mayaqq.estrogen.registry.EstrogenItems;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.packs.resources.ResourceManager;
import net.minecraft.server.packs.resources.ResourceManagerReloadListener;

import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.List;

public class ThighHighStyleLoader implements ResourceManagerReloadListener {

    public static final ThighHighStyleLoader INSTANCE = new ThighHighStyleLoader();

    public static final Codec<List<ResourceLocation>> STYLES_CODEC = Codec.list(ResourceLocation.CODEC);

    @Override
    public void onResourceManagerReload(ResourceManager resourceManager) {
        final List<ResourceLocation> styles = new ArrayList<>();
        resourceManager.getResourceStack(Estrogen.id("thigh_high_styles.json")).forEach(resource -> {
            try(Reader reader = resource.openAsReader()) {
                JsonObject root = JsonParser.parseReader(reader).getAsJsonObject();
                JsonElement replaceElement = root.get("replace");

                boolean replace = replaceElement.isJsonPrimitive() ? replaceElement.getAsBoolean() : false;
                if(replace) styles.clear();

                List<ResourceLocation> added = STYLES_CODEC.decode(JsonOps.INSTANCE, root.get("values"))
                    .resultOrPartial(Estrogen.LOGGER::error)
                    .map(Pair::getFirst)
                    .orElseGet(List::of);

                styles.addAll(added);
            } catch (IOException | JsonParseException e) {
                Estrogen.LOGGER.error("Invalid thigh high styles:", e);
            }
        });
        EstrogenItems.THIGH_HIGHS.get().loadStyles(styles);
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/features/thighhighs/ThighHighStyleLoader.kt
package dev.mayaqq.estrogen.features.thighhighs

import com.google.gson.JsonParser
import com.mojang.serialization.Codec
import com.mojang.serialization.JsonOps
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.mayaqq.cynosure.core.VersionHooks.Impl.toKtResult
import dev.mayaqq.cynosure.core.codecs.fieldOf
import dev.mayaqq.cynosure.core.codecs.forGetter
import dev.mayaqq.estrogen.Estrogen
import dev.mayaqq.estrogen.content.EstrogenItems
import dev.mayaqq.estrogen.id
import invoke.kitty.kritter.events.DataPackSyncEvent
import invoke.kitty.kritter.resources.AsyncResourceReloadListener
import invoke.kitty.kritter.utils.coroutines.mapAsync
import invoke.kitty.kritter.utils.result.flatMap
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Semaphore
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.util.profiling.ProfilerFiller


object ThighHighStyleLoader : AsyncResourceReloadListener<List<ThighHighStyleLoader.StyleInstance?>> {

    init {
        DataPackSyncEvent.subscribe { player, _ ->
            EstrogenItems.ThighHighs.get().syncStyles(player)
        }
    }

    data class StyleInstance(val replace: Boolean, val values: List<ResourceLocation>)

    val STYLES_CODEC: Codec<StyleInstance> = RecordCodecBuilder.create { instance -> instance.group(
        Codec.BOOL.optionalFieldOf("replace", false) forGetter StyleInstance::replace,
        ResourceLocation.CODEC.listOf() fieldOf StyleInstance::values
    ).apply(instance, ::StyleInstance) }

    override suspend fun load(resourceManager: ResourceManager, profiler: ProfilerFiller): List<StyleInstance?> = coroutineScope {
        resourceManager.getResourceStack(id("thigh_high_styles.json"))
            .mapAsync(Semaphore(16)) { resource ->
                resource.openAsReader().use { reader ->
                    runCatching { JsonParser.parseReader(reader) }
                        .flatMap { STYLES_CODEC.parse(JsonOps.INSTANCE, it).toKtResult() }
                        .onFailure { Estrogen.error("Error loading thigh high styles", it) }
                        .getOrNull()
                }
            }
    }

    override suspend fun apply(data: List<StyleInstance?>, resourceManager: ResourceManager, profiler: ProfilerFiller) {
        EstrogenItems.ThighHighs.get().loadStyles(
            data.filterNotNull()
                .takeLastWhile { !it.replace }
                .flatMap { it.values }
        )
    }
}